standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java

```diff
@@ -10234,93 +10234,105 @@ protected Integer getJdoResult(GetHelper<Integer> ctx) throws MetaException, NoS
+                + "t4.contains(colName) && partition.table.database.catalogName == t5" + (engine != null ? " && engine == t6" : "");
+        parameters = "java.lang.String t1, java.lang.String t2, java.lang.String t3, "
+                + "java.util.Collection t4, java.lang.String t5" + (engine != null ? ", java.lang.String t6" : "");
       } else {
         filter = "partition.partitionName == t1 && partition.table.database.name == t2 && partition.table.tableName == t3 && partition.table.database.catalogName == t4" + (engine != null ? " && engine == t5" : "");
         parameters = "java.lang.String t1, java.lang.String t2, java.lang.String t3, java.lang.String t4" + (engine != null ? ", java.lang.String t5" : "");
       }
       query.setFilter(filter);
       query.declareParameters(parameters);
-      if (colName != null) {
-        query.setUnique(true);
+      if (colNames != null) {
+        List<String> normalizedColNames = new ArrayList<>();
+        for (String colName : colNames){
+          // trim the extra spaces, and change to lowercase
+          normalizedColNames.add(normalizeIdentifier(colName));
+        }
         if (engine != null) {
-          mStatsObj =
-              (MPartitionColumnStatistics) query.executeWithArray(partName.trim(),
-                  normalizeIdentifier(dbName),
-                  normalizeIdentifier(tableName),
-                  normalizeIdentifier(colName),
-                  normalizeIdentifier(catName),
-                  engine);
+          mStatsObjColl =
+                  (List<MPartitionColumnStatistics>) query.executeWithArray(partName.trim(),
+                          normalizeIdentifier(dbName),
+                          normalizeIdentifier(tableName),
+                          normalizedColNames,
+                          normalizeIdentifier(catName),
+                          engine);
         } else {
-          mStatsObj =
-              (MPartitionColumnStatistics) query.executeWithArray(partName.trim(),
-                  normalizeIdentifier(dbName),
-                  normalizeIdentifier(tableName),
-                  normalizeIdentifier(colName),
-                  normalizeIdentifier(catName));
-        }
-        pm.retrieve(mStatsObj);
-        if (mStatsObj != null) {
-          pm.deletePersistent(mStatsObj);
+          mStatsObjColl =
+                  (List<MPartitionColumnStatistics>) query.executeWithArray(partName.trim(),
+                          normalizeIdentifier(dbName),
+                          normalizeIdentifier(tableName),
+                          normalizedColNames,
+                          normalizeIdentifier(catName));
+        }
+        pm.retrieveAll(mStatsObjColl);
+        if (mStatsObjColl != null) {
+          pm.deletePersistentAll(mStatsObjColl);
         } else {
           throw new NoSuchObjectException("Column stats doesn't exist for table="
```

Comment:
nit: this exception should never throw?

Reply:
Let's discuss this today. I am not sure if I understand this.